anoma/namada-interface
# feat/1295 - Update Namada Keychain integration

resolves #1295 
resolves #1270 

- [x] Clean up `Signer` class (keep only signing-related functions, similar to Keplr interface)
- [x] Update `namada.connect()`, `namada.disconnect()` to accept optional `chainId` - this will be used to set approved signing chain, but may not be added in this PR (depending on time) but it will be good to have to be backward compatible
- [x] Namadillo should provide current `chainId` to extension on connect, is connected, and disconnect
- [x] Remove dependency on `@namada/integrations` - this may take a few iterations, but once Namadillo is completely updated, we can probably just remove this package
- [x] Fix Namadillo issues with detection (to trigger this behavior reliably, change something in Settings)
- [x] Update Faucet app to account for any API changes (will fix Docker build)
- [x] Approve Connection should also approve `chainId` for signing (now that we have the `chainId` - may need to adjust a bit so that pointing to a new indexer (if it has a different `chainId`) triggers an approval _or_ `Connect` button
- [x] Remove  deprecated viewing keys function in `balance/atoms.ts`
  - [ ] TODO: When we make `chainId` required for approvals, update tests to reflect this. For now, it should work the old way as well, only it won't update the value in `Network`

### Notes

- FYI,  I had originally implemented a wallet manager for Namada Keychain, but I think it should only share a part of that interface since it wouldn't ever work well with the Keplr wallet manager, so I'm making a separate hook specifically for Namada Keychain, as our current integration is already separate

### TESTING

__NOTE__ In general, we just want to make sure any extension functionality works as expected, and if the extension is installed, you should never see the `Download` button. The following is how I tested this:

- From this branch, load Dry-Run settings and connect extension (you should be prompted to approve domain & enable signing for chainId:
  ```bash
  https://indexer.namada-dryrun.tududes.com
  https://rpc.namada-dryrun.tududes.com
  https://masp.namada-dryrun.tududes.com
  ```
- Check `Network` in extension, this should be set to the current chain ID
- Load a different indexer (Housefire) in Namadillo: `https://indexer.knowable.run` - you should be prompted to approve once the new chain ID is set
- Check `Network` in extension, this should be set to the new chain ID. Also, you can check this in the console with: `await namada.getChain()`
- Test sign Tx and sign-arbitrary
- Test switching users and `Disconnect`
- Run `await namada.accounts()` to see cleaned up data
- You must now use `const signer = namada.getSigner()` instead of `namada.signer()` if you want to use the Signer API (same results as the `namada.signTx()` calls, only the usage is different as before). 
- Also, test for any cases where `Download` is displayed when the extension is installed - this should be fixed now!
- In Chrome, launch a private tab to see the `Download` state (`CMD+Shift+N`)
